File: JavaRewrite/src/main/java/io/github/deadshot465/assignments/musicians/BandSupervisor.java
package io.github.deadshot465.assignments.musicians;

import java.util.HashMap;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;

public class BandSupervisor {
    private static final BlockingQueue<MusicianData> queue = new ArrayBlockingQueue<>(100);

    public static void startBand(int maxRetries) throws InterruptedException {
        final var members = new HashMap<String, Musician>();
        members.put("singer", addBandMember("singer", SkillLevel.Good));
        members.put("bass", addBandMember("bass", SkillLevel.Good));
        members.put("drum", addBandMember("drum", SkillLevel.Bad));
        members.put("guitar", addBandMember("guitar", SkillLevel.Good));

        while (true) {
            var newMusicianData = queue.poll(1000, TimeUnit.MILLISECONDS);
            if (newMusicianData == null) {
                continue;
            }

            maxRetries--;
            members.remove(newMusicianData.role());
            if (maxRetries <= 0) {
                System.out.println("The manager is mad and fired the whole band!");
                for (var musician :
                        members.values()) {
                    musician.setFired(true);
                }
                Thread.sleep(3000);
                break;
            }
            members.put(newMusicianData.role(), addBandMember(newMusicianData.role(), newMusicianData.skillLevel()));
        }
    }

    private static Musician addBandMember(String role, SkillLevel skillLevel) {
        final var musician = new Musician(role, skillLevel);
        final var thread = new Thread(() -> {
            while (true) {
                try {
                    Thread.sleep(Musician.delay);
                    final var playResult = musician.playSound();
                    if (playResult) {
                        continue;
                    }

                    queue.add(new MusicianData(role, skillLevel));
                    break;
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        });

        thread.start();
        return musician;
    }
}

File: TypeScript/assignments/Musicians/BandSupervisor.ts
import { Musician, SkillLevel } from "./Musician.ts";

const delay = 750;

let identifier = 0;

export function startBand(maxRetries: number) {
    const members = new Map<string, Musician>();
    members.set("singer", addBandMember("singer", SkillLevel.Good));
    members.set("bass", addBandMember("bass", SkillLevel.Good));
    members.set("drum", addBandMember("drum", SkillLevel.Bad));
    members.set("guitar", addBandMember("guitar", SkillLevel.Good));

    const stopInterval = () => {
        if (maxRetries <= 0) {
            clearInterval(identifier);
        }
    }

    identifier = setInterval(() => {
        for (const member of members) {
            const playResult = member[1].playSound();
            if (playResult) {
                continue;
            }

            if (maxRetries <= 0) {
                continue;
            }

            maxRetries--;
            if (maxRetries <= 0) {
                console.log("The manager is mad and fired the whole band!");
                for (const m of members) {
                    m[1].setFired();
                }
                return;
            }

            const {role, skillLevel} = member[1];
            members.delete(role);
            members.set(role, addBandMember(role, skillLevel));
        }
        stopInterval();
    }, delay);
} 

function addBandMember(role: string, skillLevel: SkillLevel): Musician {
    const musician = new Musician(role, skillLevel);
    return musician;
}